### PR TITLE
[Benchmark] add tpu and torchscipt for benchmark

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR adds:

- Torchscript memory and time benchmarking
- TPU memory and time benchmarking

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
**Baseline GPU results**:

======= INFERENCE - SPEED - RESULT =======
	======= MODEL CHECKPOINT: distilbert-base-uncased =======
		distilbert-base-uncased/8/8: 0.007s
		distilbert-base-uncased/8/32: 0.009s
		distilbert-base-uncased/8/128: 0.022s
		distilbert-base-uncased/8/512: 0.1s
	======= MODEL CHECKPOINT: bert-base-cased =======
		bert-base-cased/8/8: 0.015s
		bert-base-cased/8/32: 0.025s
		bert-base-cased/8/128: 0.072s
		bert-base-cased/8/512: 0.332s
======= INFERENCE - MEMORY - RESULT =======
	======= MODEL CHECKPOINT: distilbert-base-uncased =======
		distilbert-base-uncased/8/8: 274 MB
		distilbert-base-uncased/8/32: 298 MB
		distilbert-base-uncased/8/128: 324 MB
		distilbert-base-uncased/8/512: 552 MB
	======= MODEL CHECKPOINT: bert-base-cased =======
		bert-base-cased/8/8: 458 MB
		bert-base-cased/8/32: 462 MB
		bert-base-cased/8/128: 488 MB
		bert-base-cased/8/512: 728 MB

**Comment by patrickvonplaten** [comment]
** Torchscript GPU results:**

======= INFERENCE - SPEED - RESULT =======
	======= MODEL CHECKPOINT: distilbert-base-uncased =======
		distilbert-base-uncased/8/8: 0.005s
		distilbert-base-uncased/8/32: 0.009s
		distilbert-base-uncased/8/128: 0.02s
		distilbert-base-uncased/8/512: 0.096s
	======= MODEL CHECKPOINT: bert-base-cased =======
		bert-base-cased/8/8: 0.012s
		bert-base-cased/8/32: 0.025s
		bert-base-cased/8/128: 0.073s
		bert-base-cased/8/512: 0.328s
======= INFERENCE - MEMORY - RESULT =======
	======= MODEL CHECKPOINT: distilbert-base-uncased =======
		distilbert-base-uncased/8/8: 274 MB
		distilbert-base-uncased/8/32: 296 MB
		distilbert-base-uncased/8/128: 312 MB
		distilbert-base-uncased/8/512: 552 MB
	======= MODEL CHECKPOINT: bert-base-cased =======
		bert-base-cased/8/8: 458 MB
		bert-base-cased/8/32: 460 MB
		bert-base-cased/8/128: 488 MB
		bert-base-cased/8/512: 716 MB

check colab here: https://colab.research.google.com/drive/10KSu_6X6unsKXPOiwiGP6QDC1fLtADFJ?usp=sharing

The differences seem very small to me. What do you think @LysandreJik ?

File: src/transformers/benchmark/benchmark.py [hunk 1]
```diff
@@ -58,15 +68,25 @@ def train(self, model_name, batch_size, sequence_length, trace_memory=False):
                 vocab_size, (batch_size, sequence_length), dtype=torch.long, device=self.args.device
             )
 
+            if self.args.torchscript:
+                raise NotImplementedError("Training for torchscript is currently not implemented")
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark.py:72]: @LysandreJik I think to implement torchscript for training, I need to pass all tensors until labels. I don't think we always have the same number of tensors before `labels` in each model's forward(). So not sure what to do here...Could do something ugly with `if else statements depending on the model class`. What do you think?

File: src/transformers/benchmark/benchmark.py [hunk 2]
```diff
@@ -149,11 +182,22 @@ def inference(self, model_name, batch_size, sequence_length, trace_memory=False)
                 vocab_size, (batch_size, sequence_length), dtype=torch.long, device=self.args.device
             )
 
+            if self.args.torchscript:
+                with torch.no_grad():
+                    if config.is_encoder_decoder:
+                        raise NotImplementedError("Torchscript is currently not supported for EncoderDecoder models")
+                    else:
+                        inference_model = torch.jit.trace(model, input_ids)
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark.py:190]: This is more or less all I added for torchscript...is there anything else I should pay attention to?

**Comment by patrickvonplaten** [comment]
**TPU memory and time usage**

======= INFERENCE - SPEED - RESULT =======
	======= MODEL CHECKPOINT: distilbert-base-uncased =======
		distilbert-base-uncased/8/8: 0.004s
		distilbert-base-uncased/8/32: 0.005s
		distilbert-base-uncased/8/128: 0.004s
		distilbert-base-uncased/8/512: 0.005s
	======= MODEL CHECKPOINT: bert-base-cased =======
		bert-base-cased/8/8: 0.01s
		bert-base-cased/8/32: 0.008s
		bert-base-cased/8/128: 0.009s
		bert-base-cased/8/512: 0.009s
TPU was used for inference. Note that the time after compilation stabilized (after ~10 inferences model.forward(..) calls) was measured.
======= INFERENCE - MEMORY - RESULT =======
	======= MODEL CHECKPOINT: distilbert-base-uncased =======
		distilbert-base-uncased/8/8: 1027 MB
		distilbert-base-uncased/8/32: 1118 MB
		distilbert-base-uncased/8/128: 1118 MB
		distilbert-base-uncased/8/512: 1118 MB
	======= MODEL CHECKPOINT: bert-base-cased =======
		bert-base-cased/8/8: 1330 MB
		bert-base-cased/8/32: 1332 MB
		bert-base-cased/8/128: 1332 MB
		bert-base-cased/8/512: 1332 MB

File: src/transformers/benchmark/benchmark.py [hunk 3]
```diff
@@ -204,10 +255,27 @@ def encoder_forward():
 
                 return memory, summary
             else:
+
+                if not self.args.no_tpu and is_tpu_available():
+                    # run additional 10 times to stabilize compilation for tpu
+                    logger.info("Do inference on TPU. Running model 10 times to stabilize compilation")
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark.py:261]: @LysandreJik - For TPU I ran the function 10 times before measuring to stabilize compilation. Should I do the same for torchscript?

File: src/transformers/training_args.py [hunk 4]
```diff
@@ -5,25 +5,15 @@
...
 
 if is_torch_available():
     import torch
 
-
-try:
+if is_tpu_available():
     import torch_xla.core.xla_model as xm
 
-    _has_tpu = True
```
Comment (patrickvonplaten) [inline src/transformers/training_args.py:18]: @julien-c I moved the `_has_tpu` logic to the `file_utils.py` since it's similar to `is_torch_available()` IMO.

File: src/transformers/benchmark/benchmark_utils.py [hunk 5]
```diff
@@ -128,6 +131,127 @@ class MemorySummary(NamedTuple):
 MemoryTrace = List[UsedMemoryState]
 
 
+def measure_peak_memory_cpu(function: Callable[[], None], interval=0.5) -> int:
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark_utils.py:134]: clean function to measure peak cpu memory usage

File: setup.py [hunk 6]
```diff
@@ -84,7 +84,7 @@
 extras["serving"] = ["pydantic", "uvicorn", "fastapi", "starlette"]
 extras["all"] = extras["serving"] + ["tensorflow", "torch"]
 
-extras["testing"] = ["pytest", "pytest-xdist", "timeout-decorator"]
+extras["testing"] = ["pytest", "pytest-xdist", "timeout-decorator", "psutil"]
```
Comment (patrickvonplaten) [inline setup.py:87]: @thomwolf @LysandreJik @julien-c - Hope it's ok to add psutil for testing purposes.

File: src/transformers/training_args.py (hunk 4 above)
Comment (LysandreJik) [reply src/transformers/training_args.py:18]: I think this is cool.

**Review by LysandreJik — APPROVED** [review]
Nice! I would indeed to a few inferences for torchscript as well, so that it's traced beforehand.

**Comment by LysandreJik** [comment]
Requesting @julien-c's review for the move of `is_tpu_available` to the utils.

**Comment by patrickvonplaten** [comment]
UPDATE:

I'm fine with PyTorch for CPU, GPU with and without torchscript results:
https://docs.google.com/spreadsheets/d/1vgAIG7P3AOdBp5X91rVVu8AqnZ_hAvFzKj_fTNolAlU/edit?usp=sharing

TPU running times also seem to be fine. TPU memory is not yet implemented - will probably wait here until there is a PyTorch XLA API: https://github.com/pytorch/xla/issues/2180

**Comment by patrickvonplaten** [comment]
Good to merge for me, waiting for @julien-c to check it out

**Comment by julien-c** [comment]
Good for me

File: setup.py (hunk 6 above)
Comment (julien-c) [reply setup.py:87]: Yes that’s a robust dependency

File: src/transformers/training_args.py (hunk 4 above)
Comment (julien-c) [reply src/transformers/training_args.py:18]: Should we make explicit that it’s TPU available *in PyTorch*?

Comment (patrickvonplaten) [reply src/transformers/training_args.py:18]: Agree! Will change that to `is_torch_tpu_available()`

**Comment by patrickvonplaten** [comment]
Okey changed it to `is_torch_tpu_available()`. Think that's fine. Pinging @julien-c @LysandreJik to notice the change.

**Comment by LysandreJik** [comment]
Indeed, nice change!
